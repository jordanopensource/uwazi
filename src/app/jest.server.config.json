{
  "workerIdleMemoryLimit": 0.2,
  "displayName": "Server",
  "testMatch": ["**/app/api/**/specs/*spec.(j|t)s?(x)", "**/app/shared/**/specs/*spec.(j|t)s?(x)"],
  "testEnvironment": "node",
  "setupFilesAfterEnv": ["<rootDir>/setUpJestServer.js"],
  "snapshotFormat": {
    "escapeString": true,
    "printBasicPrototype": true
  },
  "fakeTimers": {
    "legacyFakeTimers": true
  },
  "transform": {
    "\\.[jt]sx?$": ["babel-jest", { "rootMode": "upward" }]
  },
  "moduleNameMapper": {
    "^api/(.*)": "<rootDir>/api/$1",
    "^shared/(.*)": "<rootDir>/shared/$1",
    "^uuid$": "<rootDir>/../node_modules/uuid/dist/index.js"
  }
}
