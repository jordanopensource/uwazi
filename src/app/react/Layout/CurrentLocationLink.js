import React from 'react';
import PropTypes from 'prop-types';
import { Link, useLocation } from 'react-router-dom';

const validProps = props => {
  const { to, ...valid } = props;
  return valid;
};

const CurrentLocationLink = ({ children, queryParams, ...otherProps }) => {
  const location = useLocation();
  const query = new URLSearchParams(location.search);
  Object.keys(queryParams).forEach(key => {
    query.set(key, queryParams[key]);
  });

  query.forEach((value, key) => {
    if (value === '') {
      query.delete(key);
    }
  });

  return (
    <Link to={`${location.pathname}?${query.toString()}`} {...validProps(otherProps)}>
      {children}
    </Link>
  );
};

CurrentLocationLink.defaultProps = {
  children: '',
  queryParams: {},
};

CurrentLocationLink.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  queryParams: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

export { CurrentLocationLink };

export default CurrentLocationLink;
