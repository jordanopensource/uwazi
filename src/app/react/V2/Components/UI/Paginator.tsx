import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { ChevronLeftIcon, ChevronRightIcon } from '@heroicons/react/20/solid';
import { Translate } from 'app/I18N';

interface PaginatorProps {
  currentPage: number;
  totalPages: number;
  buildUrl: (page: string) => string;
  preventScrollReset?: boolean;
}

const calculateMorePages = (currentPage: number, totalPages: number) => {
  let page = currentPage + 1;
  const pages = [];

  while (pages.length < 5) {
    if (page + 1 === totalPages) {
      break;
    }

    pages.push((page + 1).toString());
    page += 1;
  }

  return pages;
};

const Paginator = ({ currentPage, totalPages, buildUrl, preventScrollReset }: PaginatorProps) => {
  const isFirstPage = currentPage === 1;
  const isLastPage = totalPages < 1 || currentPage === totalPages;
  const shouldDisplayShowMore = currentPage + 1 !== totalPages && !isLastPage;
  const [showMore, setShowMore] = useState<Boolean>(totalPages - currentPage < 6);

  return (
    <nav aria-label="Pagination">
      <ul className="flex flex-wrap items-center">
        <li key="previous">
          {isFirstPage ? (
            <button
              type="button"
              disabled
              className="block px-3 py-1.5 ml-0 leading-snug text-gray-500 bg-white border border-gray-300 rounded-l-[4px]"
            >
              <ChevronLeftIcon className="w-[17px]" />
              <Translate className="sr-only">Previous</Translate>
            </button>
          ) : (
            <Link
              to={buildUrl((currentPage - 1).toString())}
              preventScrollReset={preventScrollReset}
              className="block px-3 py-1.5 ml-0 leading-snug text-gray-500 bg-white border border-gray-300 rounded-l-[4px] hover:bg-gray-100 hover:text-gray-700"
            >
              <Translate className="sr-only">Previous</Translate>
              <ChevronLeftIcon className="w-[17px]" />
            </Link>
          )}
        </li>

        {!isFirstPage && (
          <li key="first">
            <Link
              to={buildUrl('1')}
              preventScrollReset={preventScrollReset}
              aria-current="page"
              className="px-3 py-1.5 leading-snug text-gray-500 bg-white border-r border-gray-300 border-y hover:bg-gray-100 hover:text-gray-700"
            >
              1
            </Link>
          </li>
        )}

        {!isFirstPage && currentPage - 1 !== 1 && (
          <li key={currentPage - 1}>
            <Link
              to={buildUrl((currentPage - 1).toString())}
              preventScrollReset={preventScrollReset}
              aria-current="page"
              className="px-3 py-1.5 leading-snug text-gray-500 bg-white border-r border-gray-300 border-y hover:bg-gray-100 hover:text-gray-700"
            >
              {currentPage - 1}
            </Link>
          </li>
        )}

        <li key="current">
          <Link
            to={buildUrl(currentPage.toString())}
            preventScrollReset={preventScrollReset}
            aria-current="page"
            className="px-3 py-1.5 leading-snug text-blue-600 border-r border-blue-300 border-y bg-blue-50"
          >
            {currentPage}
          </Link>
        </li>

        {!isLastPage && currentPage + 1 !== totalPages && (
          <li key={currentPage + 1}>
            <Link
              to={buildUrl((currentPage + 1).toString())}
              preventScrollReset={preventScrollReset}
              aria-current="page"
              className="px-3 py-1.5 leading-snug text-gray-500 bg-white border-r border-gray-300 border-y hover:bg-gray-100 hover:text-gray-700"
            >
              {currentPage + 1}
            </Link>
          </li>
        )}

        {shouldDisplayShowMore &&
          (showMore ? (
            calculateMorePages(currentPage, totalPages).map(pageNumber => (
              <li key={`more-${pageNumber}`}>
                <Link
                  to={buildUrl(pageNumber)}
                  preventScrollReset={preventScrollReset}
                  className="px-3 py-1.5 leading-snug text-gray-500 bg-white border-r border-gray-300 border-y hover:bg-gray-100 hover:text-gray-700"
                >
                  {pageNumber}
                </Link>
              </li>
            ))
          ) : (
            <li key="more">
              <button
                onClick={() => setShowMore(true)}
                type="button"
                className="h-[35px] px-3 py-1.5 m-0 leading-snug text-gray-500 bg-white border-y border-r border-gray-300 hover:bg-gray-100 hover:text-gray-700"
              >
                <Translate className="sr-only">Show more</Translate>
                ...
              </button>
            </li>
          ))}

        {!isLastPage && (
          <li key="last">
            <Link
              to={buildUrl(totalPages.toString())}
              preventScrollReset={preventScrollReset}
              className="px-3 py-1.5 leading-snug text-gray-500 bg-white border-gray-300 border-y hover:bg-gray-100 hover:text-gray-700"
            >
              {totalPages}
            </Link>
          </li>
        )}

        <li key="next">
          {isLastPage ? (
            <button
              type="button"
              disabled
              className="block px-3 py-1.5 leading-snug text-gray-500 bg-white border border-gray-300 rounded-r-[4px]"
            >
              <ChevronRightIcon className="w-[17px]" />
              <Translate className="sr-only">Next</Translate>
            </button>
          ) : (
            <Link
              to={buildUrl((currentPage + 1).toString())}
              preventScrollReset={preventScrollReset}
              className="block px-3 py-1.5 leading-snug text-gray-500 bg-white border border-gray-300 rounded-r-[4px] hover:bg-gray-100 hover:text-gray-700"
            >
              <Translate className="sr-only">Next</Translate>
              <ChevronRightIcon className="w-[17px]" />
            </Link>
          )}
        </li>
      </ul>
    </nav>
  );
};

export { Paginator };
