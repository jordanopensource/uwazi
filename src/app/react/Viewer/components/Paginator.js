/* eslint-disable react/jsx-props-no-spreading */
import PropTypes from 'prop-types';
import React from 'react';
import { useSearchParams } from 'react-router-dom';
import { Translate } from 'app/I18N';
import { CurrentLocationLink } from 'app/Layout';
import { searchParamsFromSearchParams } from 'app/utils/routeHelpers';

const disableButton = (page, pageToDisable) => ({
  className: page === pageToDisable ? 'btn disabled' : 'btn',
  rel: page === pageToDisable ? 'nofollow' : undefined,
});

const Paginator = ({ page, totalPages, onPageChange }) => {
  const prevPage = page - 1 || 1;
  const nextPage = page + 1 > totalPages ? totalPages : page + 1;
  return (
    <div className="paginator">
      <CurrentLocationLink
        queryParams={{ page: prevPage }}
        onClick={e => {
          e.preventDefault();
          onPageChange(prevPage);
        }}
        {...disableButton(page, 1)}
      >
        <Translate>Previous</Translate>
      </CurrentLocationLink>
      <span>{` ${page} / ${totalPages} `}</span>
      <CurrentLocationLink
        queryParams={{ page: nextPage }}
        onClick={e => {
          e.preventDefault();
          onPageChange(nextPage);
        }}
        {...disableButton(page, totalPages)}
      >
        <Translate>Next</Translate>
      </CurrentLocationLink>
    </div>
  );
};

Paginator.defaultProps = {
  page: 1,
  totalPages: 1,
  onPageChange: () => {},
};

Paginator.propTypes = {
  page: PropTypes.number,
  totalPages: PropTypes.number,
  onPageChange: PropTypes.func,
};

// eslint-disable-next-line react/no-multi-comp
const PaginatorWithPage = props => {
  const [searchParams] = useSearchParams();
  const query = searchParamsFromSearchParams(searchParams);
  const { ...restProps } = props;
  return <Paginator {...restProps} page={Number(query.page || 1)} />;
};

export { PaginatorWithPage };
