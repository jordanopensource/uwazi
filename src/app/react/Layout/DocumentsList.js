/* eslint-disable max-lines */
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { toUrlParams } from 'shared/JSONRequest';
import rison from 'rison-node';
import { SearchBar } from 'app/Library/components/SearchBar';
import ConnectionSearchBar from 'app/ConnectionsList/components/SearchBar';
import { Loader } from 'app/components/Elements/Loader';
import Footer from 'app/App/Footer';

import { NeedAuthorization } from 'app/Auth';
import { t, Translate } from 'app/I18N';
import { DocumentCounter } from 'app/Layout/DocumentCounter';
import { Icon } from 'UI';
import { withRouter } from 'app/componentWrappers';
import { LibraryHeader } from 'app/Library/components/LibraryHeader';
import { risonDecodeOrIgnore } from 'app/utils';
import Welcome from './components/Welcome';
import { TilesViewer } from './TilesViewer';
import blankState from '../Library/helpers/blankState';

const selectAllEntities = command => {
  command.selectAllDocuments();
};
class DocumentsList extends Component {
  static getDerivedStateFromProps(newProps, prevState) {
    const { scrollCount = 0, parentScrollCount = 0 } = prevState;
    if (newProps.scrollCount !== parentScrollCount) {
      return {
        loading: false,
        scrollCount: scrollCount + 1,
        parentScrollCount: newProps.scrollCount,
      };
    }

    return { loading: false };
  }

  constructor(props, context) {
    super(props, context);
    this.state = {
      loading: false,
      scrollCount: 0,
      parentScrollCount: 0,
    };
    this.clickOnDocument = this.clickOnDocument.bind(this);
    this.selectAllDocuments = this.selectAllDocuments.bind(this);
    this.loadNextGroupOfEntities = this.loadNextGroupOfEntities.bind(this);
  }

  loadMoreDocuments(amount, from) {
    this.setState({ loading: true });
    this.props.loadMoreDocuments(amount, from, this.props.location, this.props.navigate);
  }

  selectAllDocuments() {
    if (this.props.selectAllDocuments) {
      this.props.selectAllDocuments.apply(this);
    }
  }

  clickOnDocument(...args) {
    if (this.props.clickOnDocument) {
      this.props.clickOnDocument.apply(this, args);
    }
  }

  loadNextGroupOfEntities() {
    const from = this.props.documents.get('rows').size;
    const DEFAULT_PAGE_SIZE = 30;
    if (from) {
      this.loadMoreDocuments(DEFAULT_PAGE_SIZE, from);
    }
  }

  loadMoreButton(amount) {
    const query = { ...this.props.location.query };
    const q = query.q ? risonDecodeOrIgnore(query.q) : {};
    const from = this.props.documents.get('rows').size;
    q.from = from;
    q.limit = amount;
    query.q = rison.encode(q);
    const url = `${this.props.location.pathname}${toUrlParams(query)}`;
    return (
      <Link
        to={url}
        className="btn btn-default btn-load-more"
        onClick={e => {
          e.preventDefault();
          this.loadMoreDocuments(amount, from);
        }}
      >
        {amount} {t('System', 'x more')}
      </Link>
    );
  }

  relationshipsToolbar(counter) {
    const SortButtonsRelationships = this.props.SortButtons;
    const SearchBarRelationships = this.props.SearchBar;

    return (
      <div className="relationship-toolbar">
        <div className="search-list">
          <SearchBarRelationships />
        </div>
        <div className="sort-by">
          <span className="documents-counter-sort">
            <Translate>sorted by</Translate>
          </span>
          <SortButtonsRelationships
            sortCallback={this.props.searchDocuments}
            selectedTemplates={this.props.filters.get('documentTypes')}
            stateProperty={this.props.sortButtonsStateProperty}
          />
        </div>
        <div className="documents-counter">
          <span className="documents-counter-label">{counter}</span>
        </div>
      </div>
    );
  }

  render() {
    const {
      documents,
      connections,
      GraphView,
      view,
      hideFooter,
      connectionsGroups,
      LoadMoreButton,
      rowListZoomLevel,
      CollectionViewer,
    } = this.props;

    const totalConnections = connections
      ? connectionsGroups.reduce(
          (total, g) =>
            total +
            g.get('templates').reduce((count, template) => count + template.get('count'), 0),
          0
        )
      : undefined;

    const counter = (
      <DocumentCounter
        selectedEntitiesCount={this.props.selectedDocuments.size}
        entityListCount={this.props.documents.get('rows').size}
        entityTotal={documents.get('totalRows')}
        hitsTotalRelation={documents.get('relation')}
        totalConnectionsCount={totalConnections}
      />
    );

    const FooterComponent = !hideFooter ? <Footer /> : null;

    const libraryContent = () => {
      if (view !== 'graph') {
        return (
          <CollectionViewer
            {...{
              rowListZoomLevel,
              storeKey: this.props.storeKey,
              clickOnDocument: this.clickOnDocument,
              onSnippetClick: this.props.onSnippetClick,
              deleteConnection: this.props.deleteConnection,
              loadNextGroupOfEntities: this.loadNextGroupOfEntities,
            }}
          />
        );
      }
      if (view === 'graph') {
        return <GraphView clickOnDocument={this.clickOnDocument} />;
      }
      return null;
    };
    const loadMoreSection = (
      <div className="row">
        {(() => {
          if (view !== 'graph') {
            return <p className="text-center col-sm-12 documents-counter">{counter}</p>;
          }
          return null;
        })()}
        {(() => {
          if (LoadMoreButton) {
            return <LoadMoreButton />;
          }
          if (documents.get('rows').size < documents.get('totalRows') && !this.state.loading) {
            return (
              <div className="text-center col-sm-12">
                {this.loadMoreButton(30)} {this.loadMoreButton(300)}
              </div>
            );
          }
          if (this.state.loading) {
            return <Loader />;
          }
          return null;
        })()}
        <NeedAuthorization>
          <div className="text-center col-sm-12 force-ltr protip">
            <Icon icon="lightbulb" />{' '}
            <b>
              <Translate>ProTip!</Translate>
            </b>
            <span>
              <Translate>Use</Translate>&nbsp;
              <span className="protip-key" no-translate="true">
                cmd
              </span>
              &nbsp; <Translate>or</Translate>{' '}
              <span className="protip-key" no-translate="true">
                shift
              </span>
              &nbsp; <Translate>+ click to select multiple cards.</Translate>
            </span>
          </div>
        </NeedAuthorization>
      </div>
    );

    return (
      <div className="documents-list">
        <div className="main-wrapper">
          {this.props.SortButtons === undefined && (
            <LibraryHeader
              storeKey={this.props.storeKey}
              counter={counter}
              selectAllDocuments={() => {
                selectAllEntities(this);
              }}
              SearchBar={this.props.SearchBar}
              searchCentered={this.props.searchCentered}
              filters={this.props.filters}
              tableViewMode={this.props.tableViewMode}
              scrollCount={this.state.scrollCount}
            />
          )}
          {this.props.SortButtons !== undefined && this.relationshipsToolbar(counter)}

          {blankState() && <Welcome />}

          {CollectionViewer.wrapLoader && (
            <div
              className="library-load-container"
              onScroll={() => {
                this.setState((prevState, _props) => ({
                  scrollCount: prevState.scrollCount + 1,
                }));
              }}
            >
              {libraryContent()}
              {loadMoreSection}
            </div>
          )}
          {!CollectionViewer.wrapLoader && (
            <>
              {libraryContent()}
              {loadMoreSection}
            </>
          )}

          {FooterComponent}
        </div>
      </div>
    );
  }
}

DocumentsList.defaultProps = {
  SearchBar,
  rowListZoomLevel: 0,
  CollectionViewer: TilesViewer,
  selectedDocuments: {},
  tableViewMode: false,
  scrollCount: 0,
  SortButtons: undefined,
};

DocumentsList.propTypes = {
  documents: PropTypes.object.isRequired,
  connections: PropTypes.object,
  filters: PropTypes.object,
  thesauri: PropTypes.object,
  selectedDocuments: PropTypes.instanceOf(Object),
  SearchBar: PropTypes.oneOfType([
    PropTypes.instanceOf(SearchBar),
    PropTypes.instanceOf(ConnectionSearchBar),
    PropTypes.func,
  ]),
  SortButtons: PropTypes.func,
  GraphView: PropTypes.func,
  search: PropTypes.object,
  loadMoreDocuments: PropTypes.func,
  searchDocuments: PropTypes.func,
  deleteConnection: PropTypes.func,
  sortButtonsStateProperty: PropTypes.string,
  storeKey: PropTypes.string,
  LoadMoreButton: PropTypes.func,
  onSnippetClick: PropTypes.func,
  clickOnDocument: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  selectAllDocuments: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  rowListZoomLevel: PropTypes.number,
  connectionsGroups: PropTypes.object,
  searchCentered: PropTypes.bool,
  hideFooter: PropTypes.bool,
  view: PropTypes.string,
  location: PropTypes.shape({
    pathname: PropTypes.string,
    query: PropTypes.shape({ page: PropTypes.string, raw: PropTypes.string }),
  }).isRequired,
  navigate: PropTypes.func.isRequired,
  CollectionViewer: PropTypes.func,
  tableViewMode: PropTypes.bool,
  scrollCount: PropTypes.number,
};

export { DocumentsList };

export default withRouter(DocumentsList);
