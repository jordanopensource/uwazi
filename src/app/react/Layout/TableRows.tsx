import React, { useState } from 'react';

import { connect, ConnectedProps } from 'react-redux';

import { TableRow } from 'app/Library/components/TableRow';
import { IStore, TableViewColumn } from 'app/istore';
import { EntitySchema } from 'shared/types/entityType';

interface TableRowsProps {
  columns: TableViewColumn[];
  storeKey: 'library' | 'uploads';
  clickOnDocument: (e: React.SyntheticEvent, doc: EntitySchema, active: boolean) => any;
}
const mapStateToProps = (state: IStore, props: TableRowsProps) => ({
  documents: state[props.storeKey].documents,
  selectionLength: state[props.storeKey].ui.get('selectedDocuments').size,
});

const connector = connect(mapStateToProps);
type mappedProps = ConnectedProps<typeof connector> & TableRowsProps;

const TableRowsComponent = ({
  documents,
  columns,
  clickOnDocument,
  storeKey,
  selectionLength,
}: mappedProps) => {
  const [multipleSelection, setMultipleSelection] = useState(selectionLength > 1);
  return (
    <>
      {documents.get('rows').map((entity: any) => (
        <TableRow
          entity={entity}
          columns={columns}
          clickOnDocument={clickOnDocument}
          storeKey={storeKey}
          key={entity.get('_id')}
          setMultipleSelection={setMultipleSelection}
          multipleSelection={multipleSelection}
        />
      ))}
    </>
  );
};

export const TableRows = connect(mapStateToProps)(TableRowsComponent);
