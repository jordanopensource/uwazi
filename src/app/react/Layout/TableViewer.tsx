import React, { Component } from 'react';
import { createSelector } from 'reselect';

import { connect } from 'react-redux';
import { EntityDisplayState, IStore, TableViewColumn } from 'app/istore';
import { Translate } from 'app/I18N';
import { TableRows } from 'app/Layout/TableRows';
import { CollectionViewerProps } from './CollectionViewerProps';

interface TableViewerProps extends CollectionViewerProps {
  columns: TableViewColumn[];
}

class TableViewerComponent extends Component<TableViewerProps> {
  static wrapLoader = true;

  render() {
    const columns = this.props.columns.filter((c: TableViewColumn) => !c.hidden);
    return (
      <div className="tableview-wrapper">
        <table>
          <thead>
            <tr>
              {columns.map((column: TableViewColumn, index: number) => (
                <th className={!index ? 'sticky-col' : ''} key={column.name}>
                  <div className="table-view-cell">
                    <Translate context={column.translationContext}>{column.label}</Translate>
                  </div>
                </th>
              ))}
            </tr>
          </thead>
          <tbody>
            <TableRows
              columns={columns}
              clickOnDocument={this.props.clickOnDocument}
              storeKey={this.props.storeKey}
            />
          </tbody>
        </table>
      </div>
    );
  }
}

const getTableViewColumnsSelector = (state: EntityDisplayState) => state.ui.get('tableViewColumns');
const selectTableViewColumns = createSelector(getTableViewColumnsSelector, columns =>
  columns?.toJS()
);

const mapStateToProps = (state: IStore, props: TableViewerProps) => ({
  columns: selectTableViewColumns(state[props.storeKey]),
});

export { selectTableViewColumns };
export type { TableViewerProps };

export const TableViewer = connect(mapStateToProps)(TableViewerComponent);
