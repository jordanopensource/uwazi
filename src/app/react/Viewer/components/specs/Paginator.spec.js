/**
 * @jest-environment jsdom
 */
import React from 'react';
import { CurrentLocationLink } from 'app/Layout';
import { renderConnectedMount } from 'app/utils/test/renderConnected';
import { PaginatorWithPage as Paginator } from '../Paginator';

let page = 1;
const mockUseLocation = jest.fn().mockImplementation(() => ({
  search: `?page=${page}`,
}));

const mockUseSearchParams = jest.fn().mockImplementation(() => {
  const params = new Map();
  params.set('page', page);
  return [params];
});

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useSearchParams: () => mockUseSearchParams(),
  useLocation: () => mockUseLocation(),
  // eslint-disable-next-line jsx-a11y/anchor-has-content, react/prop-types
  Link: props => <a {...props} href={props.to} />,
}));

describe('Paginator', () => {
  const render = props => renderConnectedMount(Paginator, {}, props, true);
  it('should render a previous button and next button based on the current page and total pages', () => {
    page = 5;
    const props = {
      totalPages: 25,
      baseUrl: 'url',
    };

    const component = render(props);
    expect(component).toMatchSnapshot();
  });

  describe('when base Url already has the query string "?"', () => {
    it('should add the page number properly to the query string', () => {
      page = 3;
      const props = {
        totalPages: 25,
        baseUrl: 'url?param=value',
      };

      const component = render(props);
      expect(component).toMatchSnapshot();
    });
  });

  describe('when on first page', () => {
    it('should disable the prev link', () => {
      page = 1;
      const props = {
        totalPages: 25,
        baseUrl: 'url',
      };

      const component = render(props);
      expect(component).toMatchSnapshot();
    });
  });

  describe('when on last page', () => {
    it('should disable the next link', () => {
      page = 25;
      const props = {
        totalPages: 25,
        baseUrl: 'url',
      };
      const component = render(props);
      expect(component).toMatchSnapshot();
    });
  });

  describe('when passing onPageChange callback', () => {
    it('should execute callback on prev/next passing the page selecte', () => {
      page = 5;
      const props = {
        totalPages: 25,
        baseUrl: 'url',
        onPageChange: jasmine.createSpy('onPageChange'),
      };

      const component = render(props);

      component
        .find(CurrentLocationLink)
        .at(0)
        .simulate('click', { preventDefault: () => {} });
      expect(props.onPageChange).toHaveBeenCalledWith(4);

      component
        .find(CurrentLocationLink)
        .at(1)
        .simulate('click', { preventDefault: () => {} });
      expect(props.onPageChange).toHaveBeenCalledWith(6);
    });
  });
});
