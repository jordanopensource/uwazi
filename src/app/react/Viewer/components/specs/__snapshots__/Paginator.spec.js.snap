// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator should render a previous button and next button based on the current page and total pages 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <PaginatorWithPage
    baseUrl="url"
    totalPages={25}
  >
    <Paginator
      baseUrl="url"
      onPageChange={[Function]}
      page={5}
      totalPages={25}
    >
      <div
        className="paginator"
      >
        <CurrentLocationLink
          className="btn"
          onClick={[Function]}
          queryParams={
            Object {
              "page": 4,
            }
          }
        >
          <Link
            className="btn"
            onClick={[Function]}
            to="undefined?page=4"
          >
            <a
              className="btn"
              href="undefined?page=4"
              onClick={[Function]}
              to="undefined?page=4"
            >
              <Connect(Translate)>
                <Translate
                  className=""
                  context="System"
                  edit={[Function]}
                  i18nmode={true}
                  translationKey="Previous"
                >
                  <span
                    className="translation active"
                    onClick={[Function]}
                  >
                    Previous
                  </span>
                </Translate>
              </Connect(Translate)>
            </a>
          </Link>
        </CurrentLocationLink>
        <span>
           5 / 25 
        </span>
        <CurrentLocationLink
          className="btn"
          onClick={[Function]}
          queryParams={
            Object {
              "page": 6,
            }
          }
        >
          <Link
            className="btn"
            onClick={[Function]}
            to="undefined?page=6"
          >
            <a
              className="btn"
              href="undefined?page=6"
              onClick={[Function]}
              to="undefined?page=6"
            >
              <Connect(Translate)>
                <Translate
                  className=""
                  context="System"
                  edit={[Function]}
                  i18nmode={true}
                  translationKey="Next"
                >
                  <span
                    className="translation active"
                    onClick={[Function]}
                  >
                    Next
                  </span>
                </Translate>
              </Connect(Translate)>
            </a>
          </Link>
        </CurrentLocationLink>
      </div>
    </Paginator>
  </PaginatorWithPage>
</Provider>
`;

exports[`Paginator when base Url already has the query string "?" should add the page number properly to the query string 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <PaginatorWithPage
    baseUrl="url?param=value"
    totalPages={25}
  >
    <Paginator
      baseUrl="url?param=value"
      onPageChange={[Function]}
      page={3}
      totalPages={25}
    >
      <div
        className="paginator"
      >
        <CurrentLocationLink
          className="btn"
          onClick={[Function]}
          queryParams={
            Object {
              "page": 2,
            }
          }
        >
          <Link
            className="btn"
            onClick={[Function]}
            to="undefined?page=2"
          >
            <a
              className="btn"
              href="undefined?page=2"
              onClick={[Function]}
              to="undefined?page=2"
            >
              <Connect(Translate)>
                <Translate
                  className=""
                  context="System"
                  edit={[Function]}
                  i18nmode={true}
                  translationKey="Previous"
                >
                  <span
                    className="translation active"
                    onClick={[Function]}
                  >
                    Previous
                  </span>
                </Translate>
              </Connect(Translate)>
            </a>
          </Link>
        </CurrentLocationLink>
        <span>
           3 / 25 
        </span>
        <CurrentLocationLink
          className="btn"
          onClick={[Function]}
          queryParams={
            Object {
              "page": 4,
            }
          }
        >
          <Link
            className="btn"
            onClick={[Function]}
            to="undefined?page=4"
          >
            <a
              className="btn"
              href="undefined?page=4"
              onClick={[Function]}
              to="undefined?page=4"
            >
              <Connect(Translate)>
                <Translate
                  className=""
                  context="System"
                  edit={[Function]}
                  i18nmode={true}
                  translationKey="Next"
                >
                  <span
                    className="translation active"
                    onClick={[Function]}
                  >
                    Next
                  </span>
                </Translate>
              </Connect(Translate)>
            </a>
          </Link>
        </CurrentLocationLink>
      </div>
    </Paginator>
  </PaginatorWithPage>
</Provider>
`;

exports[`Paginator when on first page should disable the prev link 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <PaginatorWithPage
    baseUrl="url"
    totalPages={25}
  >
    <Paginator
      baseUrl="url"
      onPageChange={[Function]}
      page={1}
      totalPages={25}
    >
      <div
        className="paginator"
      >
        <CurrentLocationLink
          className="btn disabled"
          onClick={[Function]}
          queryParams={
            Object {
              "page": 1,
            }
          }
          rel="nofollow"
        >
          <Link
            className="btn disabled"
            onClick={[Function]}
            rel="nofollow"
            to="undefined?page=1"
          >
            <a
              className="btn disabled"
              href="undefined?page=1"
              onClick={[Function]}
              rel="nofollow"
              to="undefined?page=1"
            >
              <Connect(Translate)>
                <Translate
                  className=""
                  context="System"
                  edit={[Function]}
                  i18nmode={true}
                  translationKey="Previous"
                >
                  <span
                    className="translation active"
                    onClick={[Function]}
                  >
                    Previous
                  </span>
                </Translate>
              </Connect(Translate)>
            </a>
          </Link>
        </CurrentLocationLink>
        <span>
           1 / 25 
        </span>
        <CurrentLocationLink
          className="btn"
          onClick={[Function]}
          queryParams={
            Object {
              "page": 2,
            }
          }
        >
          <Link
            className="btn"
            onClick={[Function]}
            to="undefined?page=2"
          >
            <a
              className="btn"
              href="undefined?page=2"
              onClick={[Function]}
              to="undefined?page=2"
            >
              <Connect(Translate)>
                <Translate
                  className=""
                  context="System"
                  edit={[Function]}
                  i18nmode={true}
                  translationKey="Next"
                >
                  <span
                    className="translation active"
                    onClick={[Function]}
                  >
                    Next
                  </span>
                </Translate>
              </Connect(Translate)>
            </a>
          </Link>
        </CurrentLocationLink>
      </div>
    </Paginator>
  </PaginatorWithPage>
</Provider>
`;

exports[`Paginator when on last page should disable the next link 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <PaginatorWithPage
    baseUrl="url"
    totalPages={25}
  >
    <Paginator
      baseUrl="url"
      onPageChange={[Function]}
      page={25}
      totalPages={25}
    >
      <div
        className="paginator"
      >
        <CurrentLocationLink
          className="btn"
          onClick={[Function]}
          queryParams={
            Object {
              "page": 24,
            }
          }
        >
          <Link
            className="btn"
            onClick={[Function]}
            to="undefined?page=24"
          >
            <a
              className="btn"
              href="undefined?page=24"
              onClick={[Function]}
              to="undefined?page=24"
            >
              <Connect(Translate)>
                <Translate
                  className=""
                  context="System"
                  edit={[Function]}
                  i18nmode={true}
                  translationKey="Previous"
                >
                  <span
                    className="translation active"
                    onClick={[Function]}
                  >
                    Previous
                  </span>
                </Translate>
              </Connect(Translate)>
            </a>
          </Link>
        </CurrentLocationLink>
        <span>
           25 / 25 
        </span>
        <CurrentLocationLink
          className="btn disabled"
          onClick={[Function]}
          queryParams={
            Object {
              "page": 25,
            }
          }
          rel="nofollow"
        >
          <Link
            className="btn disabled"
            onClick={[Function]}
            rel="nofollow"
            to="undefined?page=25"
          >
            <a
              className="btn disabled"
              href="undefined?page=25"
              onClick={[Function]}
              rel="nofollow"
              to="undefined?page=25"
            >
              <Connect(Translate)>
                <Translate
                  className=""
                  context="System"
                  edit={[Function]}
                  i18nmode={true}
                  translationKey="Next"
                >
                  <span
                    className="translation active"
                    onClick={[Function]}
                  >
                    Next
                  </span>
                </Translate>
              </Connect(Translate)>
            </a>
          </Link>
        </CurrentLocationLink>
      </div>
    </Paginator>
  </PaginatorWithPage>
</Provider>
`;
